### PR TITLE
Added methods to get and set thp feature values

### PR DESCRIPTION
**Pull request by santhosh93**
These new methods added in memory.py are helpful in getting
and setting the transparent huge page feature values.

Signed-off-by: Santhosh G <santhog4@linux.vnet.ibm.com>

### CODE REVIEW

File: avocado/utils/memory.py [hunk 1]
```diff
@@ -278,3 +279,35 @@ def get_buddy_info(chunk_sizes, nodes="all", zones="all"):
...
+
+    :param value : Value to be set to feature
+    :type value : str
+    :param feature : Thp feature to set
+    :type feature : str
+    """
+    thp_path = '/sys/kernel/mm/transparent_hugepage/'
+    thp_feature_to_set = os.path.join(thp_path, feature)
+    cmd = 'echo %s > %s ' % (value, thp_feature_to_set)
+    process.system(cmd, shell=True)
```
Comment (PraveenPenguin) [inline avocado/utils/memory.py]: please use `ignore_status=True` as a pram in process.system if command fails and script will error out ?

also can we have check for if value is already set that time i think no need to run this command 
may be 
```

if value!=get_thp_value(feature):
            process.system(cmd,ignore_status=True, shell=True)
```

File: avocado/utils/memory.py [hunk 2]
```diff
@@ -278,3 +279,35 @@ def get_buddy_info(chunk_sizes, nodes="all", zones="all"):
...
+    """
+    Gets the value of the thp feature arg passed
+
+    :param feature : Thp feature to get value
+    :type feature : str
+    """
+    thp_path = '/sys/kernel/mm/transparent_hugepage/'
+    thp_feature_to_get = os.path.join(thp_path, feature)
+    cmd = 'cat %s' % thp_feature_to_get
+    value = process.system_output(cmd, shell=True)
```
Comment (narasimhan-v) [inline avocado/utils/memory.py]:
```
from avocado.utils import genio

value = genio.read_file(thp_feature_to_get)
```

File: avocado/utils/memory.py [hunk 3]
```diff
@@ -278,3 +279,35 @@ def get_buddy_info(chunk_sizes, nodes="all", zones="all"):
             buddyinfo_dict[chunk_size] += int(chunk_info)
 
     return buddyinfo_dict
+
+
+def set_thp_value(value, feature):
```
Comment (narasimhan-v) [inline avocado/utils/memory.py]: would it not be better if the order be like `(feature, value)` as we generally denote key first, and value next ?

File: avocado/utils/memory.py [hunk 4]
```diff
@@ -278,3 +279,35 @@ def get_buddy_info(chunk_sizes, nodes="all", zones="all"):
...
+    Sets THP feature to a given value
+
+    :param value : Value to be set to feature
+    :type value : str
+    :param feature : Thp feature to set
+    :type feature : str
+    """
+    thp_path = '/sys/kernel/mm/transparent_hugepage/'
+    thp_feature_to_set = os.path.join(thp_path, feature)
+    cmd = 'echo %s > %s ' % (value, thp_feature_to_set)
```
Comment (santhosh93) [inline avocado/utils/memory.py]: If invalid arguments are passed from test case like,
memory.set_thp_value('always', 'khugepaged/defrag')
Then, it gives out error like, ' IOError: [Errno 22] Invalid argument '
But if process.system is used we can get error message directly that the command has failed like below,
CmdError: Command 'echo always > /sys/kernel/mm/transparent_hugepage/khugepaged/defrag ' failed (rc=1)
So, which is the better way to indicate as a failure??

File: avocado/utils/memory.py (hunk 1 above)
Comment (santhosh93) [inline avocado/utils/memory.py]: and if we use ignore_status=True as a parameter ....Then, the test using it needs to handle if the value is not set.
So, is that ok to be checked again in the test code?

Comment (PraveenPenguin) [inline avocado/utils/memory.py]: i think so, it is user need to validate

File: avocado/utils/memory.py (hunk 3 above)
Comment (santhosh93) [inline avocado/utils/memory.py]: It can be done . Thanks..

File: avocado/utils/memory.py (hunk 4 above)
Comment (narasimhan-v) [inline avocado/utils/memory.py]: When a python implementation is already existing, why use bash ?

File: avocado/utils/memory.py (hunk 1 above)
Comment (narasimhan-v) [inline avocado/utils/memory.py]: As much as possible, avoid `ERRORing` out.

File: avocado/utils/memory.py (hunk 3 above)
Comment (ldoktor) [inline avocado/utils/memory.py]: Yep, I'd like that as well...

File: avocado/utils/memory.py (hunk 4 above)
Comment (ldoktor) [inline avocado/utils/memory.py]: Btw feel free to add `safe_write_file` which reports `AssertError` (or `RuntimeError`) to the `genio` library and use it here. I see the need for it on many places...

File: avocado/utils/memory.py (hunk 1 above)
Comment (ldoktor) [inline avocado/utils/memory.py]: Well if we failed to set the value, then it should raise an exception, ideally one type of exception like `AssertError` or `RuntimeError`, which can be caught in test. Also it's nice to list the possible exception (at least the most common ones) in docstring...

File: avocado/utils/memory.py [hunk 5]
```diff
@@ -278,3 +279,35 @@ def get_buddy_info(chunk_sizes, nodes="all", zones="all"):
...
+    Gets the value of the thp feature arg passed
+
+    :param feature : Thp feature to get value
+    :type feature : str
+    """
+    thp_path = '/sys/kernel/mm/transparent_hugepage/'
+    thp_feature_to_get = os.path.join(thp_path, feature)
+    cmd = 'cat %s' % thp_feature_to_get
+    value = process.system_output(cmd, shell=True)
+    if feature == 'enabled' or feature == 'defrag':
```
Comment (ldoktor) [inline avocado/utils/memory.py]: How about `if feature in ("enabled", "defrag"):` and please also include `shmem_enabled`, which is also of type enum.

**Comment by ldoktor** [comment]
Hello @santhosh93, is everything clear here, or do you wait for our input?

**Comment by santhosh93** [comment]
@ldoktor  :sorry for the late response...I am clearhere and working on it...Will send a new PR with changes.

File: avocado/utils/memory.py (hunk 4 above)
Comment (santhosh93) [inline avocado/utils/memory.py]: is it required to add a new method safe_write_file or can we just modify the write_file itself ??

Comment (santhosh93) [inline avocado/utils/memory.py]: If invalid data or file is passed suppose to a function safe_file_write ,
we cannot catch the AssertionError as 
file_obj.write(data) will return None in both pass or fail scenarios
So IOError happens again.
So, Please let know if  it is  ok to raise normal Exception.

Comment (ldoktor) [inline avocado/utils/memory.py]: You can use `os.write` which reports `Invalid argument`. And to your previous question I'd prefer `safe_write_file`, but adding an argument to write file `write_file(file, value, safe=False)` would be also acceptable.

**Comment by santhosh93** [comment]
Opened a new PR https://github.com/avocado-framework/avocado/pull/1785

**Pull request closed without merging**